src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -0,0 +1,673 @@
+
+                foreach (var objectProperty in syntax.Properties)
+                {
+                    var propertyType = VisitAndReturnType(objectProperty);
+                    CollectErrors(errors, propertyType);
+                }
+
+                if (errors.Any())
+                {
+                    return new ErrorTypeSymbol(errors);
+                }
+
+                // type results are cached
+                var properties = syntax.Properties
+                    .GroupBy(p => p.GetKeyText(), LanguageConstants.IdentifierComparer)
+                    .Select(group => new TypeProperty(group.Key, UnionType.Create(group.Select(p => assignedTypes[p]))));
+
+                // TODO: Add structural naming?
+                return new NamedObjectType(LanguageConstants.Object.Name, properties, additionalPropertiesType: null);
+            });
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+            => AssignTypeWithCaching(syntax, () => {
+                Visit(syntax.Value);
+
+                return assignedTypes[syntax.Value];
+            });
+
+        public override void VisitArrayItemSyntax(ArrayItemSyntax syntax)
+            => AssignTypeWithCaching(syntax, () => {
+                Visit(syntax.Value);
+
+                return assignedTypes[syntax.Value];
+            });
+
+        public override void VisitParenthesizedExpressionSyntax(ParenthesizedExpressionSyntax syntax)
+            => AssignTypeWithCaching(syntax, () => {
+                Visit(syntax.Expression);
+
+                return assignedTypes[syntax.Expression];
+            });
+
+        public override void VisitFunctionArgumentSyntax(FunctionArgumentSyntax syntax)
+            => AssignTypeWithCaching(syntax, () => {
+                Visit(syntax.Expression);
+
+                return assignedTypes[syntax.Expression];
+            });
+
+        public override void VisitArraySyntax(ArraySyntax syntax)
+            => AssignTypeWithCaching(syntax, () => {
+                var errors = new List<ErrorDiagnostic>();
+
+                var itemTypes = new List<TypeSymbol>(syntax.Children.Length);
+                foreach (var arrayItem in syntax.Children)
+                {
+                    var itemType = VisitAndReturnType(arrayItem);
+                    itemTypes.Add(itemType);
+                    CollectErrors(errors, itemType);
+                }
+
+                if (errors.Any())
+                {
+                    return new ErrorTypeSymbol(errors);
+                }
+
+                var aggregatedItemType = UnionType.Create(itemTypes);
+                if (aggregatedItemType.TypeKind == TypeKind.Union || aggregatedItemType.TypeKind == TypeKind.Never)
+                {
+                    // array contains a mix of item types or is empty
+                    // assume array of any for now
+                    return LanguageConstants.Array;
```

Comment:
>Array [](start = 45, length = 5)

Note for future: I feel we'd need to support literal array types to do this correctly.